### PR TITLE
[vSphere][virtualmachine] Add metrics to virtualmachine metricset

### PR DESCRIPTION
**Pull request by ishleenk17**
- Enhancement

This PR adds the below metrics to the VirtualMachine 

Metrics | API | Field Mappings
-- | -- | --
VirtualMachine.Summary.QuickStats.UptimeSeconds | Summary | virtualmachine.uptime
VirtualMachine.Summary.OverallStatus | Summary | virtualmachine.status
VirtualMachine.Summary.Datastore| Summary | virtualmachine.datastore.names

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

- Relates https://github.com/elastic/obs-integration-team/issues/32

### CODE REVIEW

**Comment by shmsr** [comment]
@ishleenk17 Here's the diff patch:

```diff
diff --git a/metricbeat/module/vsphere/virtualmachine/data.go b/metricbeat/module/vsphere/virtualmachine/data.go
index 7a7ae1eefd..8b41667d83 100644
--- a/metricbeat/module/vsphere/virtualmachine/data.go
+++ b/metricbeat/module/vsphere/virtualmachine/data.go
@@ -22,32 +22,13 @@ import (
 )
 
 func (m *MetricSet) eventMapping(data VMData) mapstr.M {
-
-	var (
-		freeCPU    int32
-		freeMemory int64
-	)
-
 	usedMemory := int64(data.VM.Summary.QuickStats.GuestMemoryUsage) * 1024 * 1024
 	usedCPU := data.VM.Summary.QuickStats.OverallCpuUsage
 	totalCPU := data.VM.Summary.Config.CpuReservation
 	totalMemory := int64(data.VM.Summary.Config.MemorySizeMB) * 1024 * 1024
 
-	if totalCPU > 0 {
-		freeCPU = totalCPU - usedCPU
-		// Avoid negative values if reported used CPU is slightly over total configured.
-		if freeCPU < 0 {
-			freeCPU = 0
-		}
-	}
-
-	if totalMemory > 0 {
-		freeMemory = totalMemory - usedMemory
-		// Avoid negative values if reported used memory is slightly over total configured.
-		if freeMemory < 0 {
-			freeMemory = 0
-		}
-	}
+	freeCPU := max(0, totalCPU-usedCPU)
+	freeMemory := max(0, totalMemory-usedMemory)
 
 	event := mapstr.M{
 		"name":          data.VM.Summary.Config.Name,
@@ -74,6 +55,7 @@ func (m *MetricSet) eventMapping(data VMData) mapstr.M {
 			},
 		},
 	}
+
 	if len(data.CustomFields) > 0 {
 		event["custom_fields"] = data.CustomFields
 	}
@@ -83,5 +65,13 @@ func (m *MetricSet) eventMapping(data VMData) mapstr.M {
 	if len(data.DatastoreNames) > 0 {
 		event["datastore.names"] = data.DatastoreNames
 	}
+
 	return event
 }
+
+func max[T ~int32 | ~int64](a T, b T) T {
+	if a > b {
+		return a
+	}
+	return b
+}
diff --git a/metricbeat/module/vsphere/virtualmachine/virtualmachine.go b/metricbeat/module/vsphere/virtualmachine/virtualmachine.go
index 331ee1ad07..3dff84b288 100644
--- a/metricbeat/module/vsphere/virtualmachine/virtualmachine.go
+++ b/metricbeat/module/vsphere/virtualmachine/virtualmachine.go
@@ -89,38 +89,38 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
 
 	client, err := govmomi.NewClient(ctx, m.HostURL, m.Insecure)
 	if err != nil {
-		return fmt.Errorf("error in NewClient: %w", err)
+		return fmt.Errorf("virtualmachine: error in NewClient: %w", err)
 	}
 
 	defer func() {
 		if err := client.Logout(ctx); err != nil {
-			m.Logger().Debug(fmt.Errorf("error trying to logout from vshphere: %w", err))
+			m.Logger().Debugf("Error logging out from vsphere: %v", err)
 		}
 	}()
 
 	c := client.Client
 
-	// Get custom fields (attributes) names if get_custom_fields is true.
+	// Get custom fields if enabled
 	customFieldsMap := make(map[int32]string)
 	if m.GetCustomFields {
 		var err error
 		customFieldsMap, err = setCustomFieldsMap(ctx, c)
 		if err != nil {
-			return fmt.Errorf("error in setCustomFieldsMap: %w", err)
+			return fmt.Errorf("virtualmachine: error in setCustomFieldsMap: %w", err)
 		}
 	}
 
 	// Create view of VirtualMachine objects
 	mgr := view.NewManager(c)
-
 	v, err := mgr.CreateContainerView(ctx, c.ServiceContent.RootFolder, []string{"VirtualMachine"}, true)
 	if err != nil {
-		return fmt.Errorf("error in CreateContainerView: %w", err)
+		return fmt.Errorf("virtualmachine: error in CreateContainerView: %w", err)
 	}
 
+	// Ensure view is destroyed on function exit
 	defer func() {
 		if err := v.Destroy(ctx); err != nil {
-			m.Logger().Debug(fmt.Errorf("error trying to destroy view from vshphere: %w", err))
+			m.Logger().Debugf("Error destroying view from vsphere: %v", err)
 		}
 	}()
 
@@ -128,7 +128,7 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
 	var vmt []mo.VirtualMachine
 	err = v.Retrieve(ctx, []string{"VirtualMachine"}, []string{"summary", "datastore"}, &vmt)
 	if err != nil {
-		return fmt.Errorf("error in Retrieve: %w", err)
+		return fmt.Errorf("virtualmachine: error in Retrieve: %w", err)
 	}
 
 	pc := property.DefaultCollector(c)
@@ -137,13 +137,14 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
 		var networkNames, datastoreNames []string
 		var customFields mapstr.M
 
+		// Get host information
 		if host := vm.Summary.Runtime.Host; host != nil {
 			hostID = host.Value
 			hostSystem, err := getHostSystem(ctx, c, host.Reference())
 			if err == nil {
 				hostName = hostSystem.Summary.Config.Name
 			} else {
-				m.Logger().Debug(err.Error())
+				m.Logger().Debugf("Error getting host system: %v", err)
 			}
 		} else {
 			m.Logger().Debug("'Host', 'Runtime' or 'Summary' data not found. This is either a parsing error " +
@@ -161,10 +162,11 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
 				"information returned from host/guest")
 		}
 
+		// Get network names
 		if vm.Summary.Vm != nil {
 			networkNames, err = getNetworkNames(ctx, c, vm.Summary.Vm.Reference())
 			if err != nil {
-				m.Logger().Debug(err.Error())
+				m.Logger().Debugf("Error getting network names: %v", err)
 			}
 		}
 
@@ -175,7 +177,7 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
 			if err == nil {
 				datastoreNames = append(datastoreNames, ds.Name)
 			} else {
-				m.Logger().Debug(fmt.Sprintf("error retrieving datastore name for VM %s: %v", vm.Summary.Config.Name, err))
+				m.Logger().Debugf("Error retrieving datastore name for VM %s: %v", vm.Summary.Config.Name, err)
 			}
 		}
 
@@ -218,8 +220,7 @@ func getNetworkNames(ctx context.Context, c *vim25.Client, ref types.ManagedObje
 	pc := property.DefaultCollector(c)
 
 	var vm mo.VirtualMachine
-	err := pc.RetrieveOne(ctx, ref, []string{"network"}, &vm)
-	if err != nil {
+	if err := pc.RetrieveOne(ctx, ref, []string{"network"}, &vm); err != nil {
 		return nil, fmt.Errorf("error retrieving virtual machine information: %w", err)
 	}
 
@@ -227,8 +228,11 @@ func getNetworkNames(ctx context.Context, c *vim25.Client, ref types.ManagedObje
 		return nil, errors.New("no networks found")
 	}
 
-	var networkRefs []types.ManagedObjectReference
+	networkRefs := make([]types.ManagedObjectReference, 0, len(vm.Network))
 	for _, obj := range vm.Network {
+		// NOTE(SS): See: https://github.com/KubeOperator/KubeOperator/blob/master/pkg/cloud_provider/client/vsphere.go#L135
+		// To get the network names obj.Type == "DistributedVirtualPortgroup" is also considered. Please
+		// check if it is required!
 		if obj.Type == "Network" {
 			networkRefs = append(networkRefs, obj)
 		}
@@ -236,18 +240,17 @@ func getNetworkNames(ctx context.Context, c *vim25.Client, ref types.ManagedObje
 
 	// If only "Distributed port group" was found, for example.
 	if len(networkRefs) == 0 {
-		return nil, errors.New("no networks found")
+		return nil, errors.New("no supported networks found")
 	}
 
 	var nets []mo.Network
-	err = pc.Retrieve(ctx, networkRefs, []string{"name"}, &nets)
-	if err != nil {
+	if err := pc.Retrieve(ctx, networkRefs, []string{"name"}, &nets); err != nil {
 		return nil, fmt.Errorf("error retrieving network from virtual machine: %w", err)
 	}
-	outputNetworkNames := make([]string, 0, len(nets))
-	for _, net := range nets {
-		name := strings.Replace(net.Name, ".", "_", -1)
-		outputNetworkNames = append(outputNetworkNames, name)
+
+	outputNetworkNames := make([]string, len(nets))
+	for i, net := range nets {
+		outputNetworkNames[i] = strings.ReplaceAll(net.Name, ".", "_")
 	}
 
 	return outputNetworkNames, nil
```

There is one comment I left in the code itself that needs to be addressed.

**Comment by ishleenk17** [comment]
@shmsr : Regarding DistributedVirtualPortgroup object type for Network names, yes we should ideally support this.
We will have to add this support for all metricsets.
So I have created a separate task in the [Meta](https://github.com/elastic/observability-dev/issues/3797). 
Will use that to address it for all existing metricsets.

**Review by cmacknz — APPROVED** [review]
Approving go.mod changes.

File: metricbeat/module/vsphere/virtualmachine/_meta/fields.yml [hunk 1]
```diff
@@ -61,3 +61,17 @@
       type: keyword
       description: >
         Network names
+    - name: status
+      type: keyword
+      description: >
+        Overall health and status of a virtual machine
```
Comment (niraj-elastic) [inline metricbeat/module/vsphere/virtualmachine/_meta/fields.yml:67]: Please add "." after all the descriptions.

File: metricbeat/module/vsphere/virtualmachine/_meta/fields.yml [hunk 2]
```diff
@@ -61,3 +61,17 @@
...
+    - name: uptime
+      type: long
+      description: >
+        The uptime of the VM in seconds
+    - name: datastore.names
```
Comment (niraj-elastic) [inline metricbeat/module/vsphere/virtualmachine/_meta/fields.yml:72]: We can also add datastore.count as well.

File: metricbeat/module/vsphere/virtualmachine/data_test.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+			"customField1": "value1",
+			"customField2": "value2",
+		},
+	}
+
+	event := m.eventMapping(data)
+
+	// Test CPU values
+	cpuUsed, _ := event.GetValue("cpu.used.mhz")
+	assert.EqualValues(t, 30, cpuUsed)
```
Comment (niraj-elastic) [inline metricbeat/module/vsphere/virtualmachine/data_test.go:67]: If we are hardcoding all the values than we can follow this [approach ](https://github.com/elastic/beats/pull/40536/files#diff-639473420d275e860672e2c133ca092686ce3942ad50b01ca49d2b1923d0ef5eR70)which can reduce the use of assert functions.

**Review by niraj-elastic — CHANGES_REQUESTED** [review]

**Review by shmsr — APPROVED** [review]

File: metricbeat/module/vsphere/virtualmachine/_meta/fields.yml [hunk 4]
```diff
@@ -7,57 +7,71 @@
...
-        Free Memory of Guest in bytes
+        Free Memory of Guest in bytes.
       format: bytes
     - name: custom_fields
       type: object
       object_type: keyword
       description: >
-        Custom fields
+        Custom fields.
     - name: network_names
```
Comment (shmsr) [inline metricbeat/module/vsphere/virtualmachine/_meta/fields.yml:60]: Also add network.names just like the other PR? https://github.com/elastic/beats/pull/40429#issuecomment-2303621339

File: metricbeat/module/vsphere/virtualmachine/_meta/fields.yml [hunk 5]
```diff
@@ -7,57 +7,71 @@
     - name: host.id
       type: keyword
       description: >
-        Host id
+        Host id.
     - name: host.hostname
       type: keyword
       description: >
-        Host name of the host
+        Host name of the host.
```
Comment (shmsr) [inline metricbeat/module/vsphere/virtualmachine/_meta/fields.yml:14]:
```suggestion
        Hostname of the host.
```

File: metricbeat/module/vsphere/virtualmachine/data.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+		event["datastore.names"] = data.DatastoreNames
+	}
+	return event
+}
+func max[T constraints.Ordered](a T, b T) T {
+	if a > b {
+		return a
+	}
+	return b
+}
```
[the remark this answers is not in the document]
Comment (shmsr) [reply metricbeat/module/vsphere/virtualmachine/data.go:77]: This is not cmp package. Even if it were, implementation is the same more or less everywhere (no implementation is buggy (even if experimental). Here, `max` is a builtin so need to import any library. Just removing this function should be fine.

File: metricbeat/module/vsphere/virtualmachine/data.go [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+package virtualmachine
+
+import (
+	"github.com/elastic/elastic-agent-libs/mapstr"
+)
+
+func (m *MetricSet) eventMapping(data VMData) mapstr.M {
+
+	usedMemory := int64(data.VM.Summary.QuickStats.GuestMemoryUsage) * 1024 * 1024
```
Comment (niraj-elastic) [inline metricbeat/module/vsphere/virtualmachine/data.go:26]: Can we create variable for bytes conversion. https://github.com/elastic/beats/pull/40441#:~:text=%2B%09const%20bytesMultiplier%20%3D%201000

File: metricbeat/module/vsphere/virtualmachine/virtualmachine.go [hunk 8]
```diff
@@ -106,72 +115,33 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
...
 		}
 	}()
 
 	// Retrieve summary property for all machines
 	var vmt []mo.VirtualMachine
-	err = v.Retrieve(ctx, []string{"VirtualMachine"}, []string{"summary"}, &vmt)
+	err = v.Retrieve(ctx, []string{"VirtualMachine"}, []string{"summary", "datastore"}, &vmt)
 	if err != nil {
-		return fmt.Errorf("error in Retrieve: %w", err)
+		return fmt.Errorf("virtualmachine:error in Retrieve: %w", err)
```
Comment (niraj-elastic) [inline metricbeat/module/vsphere/virtualmachine/virtualmachine.go:131]:
```suggestion
		return fmt.Errorf("virtualmachine: error in Retrieve: %w", err)
```

**Review by niraj-elastic — CHANGES_REQUESTED** [review]

File: metricbeat/module/vsphere/virtualmachine/_meta/data.json [hunk 9]
```diff
@@ -15,35 +15,58 @@
...
     "vsphere": {
         "virtualmachine": {
-            "cpu": {
-                "used": {
-                    "mhz": 0
-                }
+          "host.id": "host-30",
+          "datastore.count": 1,
+          "name": "PHX-NETLAB_to_PHX-VCF-IX-R1",
+          "network.count": 3,
```
[the remark this answers is not in the document]
Comment (ishleenk17) [reply metricbeat/module/vsphere/virtualmachine/_meta/data.json:21]: Do you mean the count numbers ?

[the remark this answers is not in the document]
Comment (ishleenk17) [reply metricbeat/module/vsphere/virtualmachine/_meta/data.json]: You meant the nesting. Taken care of now.

**Review by niraj-elastic — APPROVED** [review]

**Pull request merged**